### PR TITLE
[v1.x] Restore quantized RNN operator from MXNet 1.6

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
This PR is backport of this change: https://github.com/apache/incubator-mxnet/pull/18028
with additional refactor & bug fixes from current v1.x branch

Connected to issue: https://github.com/apache/incubator-mxnet/issues/20737

@ssriram1978
@josephevans 
Test are passing, however feedback if everything works fine in your model would be appreciated

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [website]

File: src/operator/quantization/quantize_graph_pass.cc [hunk 1]
```diff
@@ -314,9 +326,20 @@ Graph QuantizeGraph(Graph &&src) {
               }
             }
 
-            ObjectPtr quantize_node = InsertNode("_contrib_quantize_v2",
-              e.node->attrs.name + suffix + "_quantize", new_node, mirror_entry);
-            quantize_node->attrs.dict["out_type"] = quantized_dtype;
+            // ObjectPtr quantize_node = InsertNode("_contrib_quantize_v2",
```
Comment (anko-intel) [inline src/operator/quantization/quantize_graph_pass.cc:329]: old code left intentionally ?

Comment (bgawrych) [reply src/operator/quantization/quantize_graph_pass.cc:329]: Nope, thanks!

**Review by josephevans — APPROVED** [review]
Thanks for porting this - added comments about the "Copyright by Contributors" lines (they should not be in the code.)

File: src/operator/quantization/mkldnn/mkldnn_quantized_rnn-inl.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2020 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/mkldnn/mkldnn_quantized_rnn-inl.h:21]: These copyright lines should not be in ASF-licensed files.

File: src/operator/quantization/mkldnn/mkldnn_quantized_rnn.cc [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2020 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/mkldnn/mkldnn_quantized_rnn.cc:21]: These copyright lines should not be in ASF-licensed files.

File: src/operator/quantization/quantize_asym-inl.h [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2020 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/quantize_asym-inl.h:21]: These copyright lines should not be in ASF-licensed files.

File: src/operator/quantization/quantize_asym.cc [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2020 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/quantize_asym.cc:21]: These copyright lines should not be in ASF-licensed files.

File: src/operator/quantization/quantize_v2.cc [hunk 6]
```diff
@@ -19,7 +19,7 @@
 
 /*!
  *  Copyright (c) 2017 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/quantize_v2.cc:21]: These copyright lines should not be in ASF-licensed files.

File: src/operator/quantization/quantized_rnn-inl.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2020 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/quantized_rnn-inl.h:21]: These copyright lines should not be in ASF-licensed files.

File: src/operator/quantization/quantized_rnn.cc [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2020 by Contributors
```
Comment (josephevans) [inline src/operator/quantization/quantized_rnn.cc:21]: These copyright lines should not be in ASF-licensed files.

File: include/mxnet/op_attr_types.h [hunk 9]
```diff
@@ -348,6 +348,21 @@ using FAvoidQuantizeInput = std::function<bool (const NodeAttrs& attrs,
                                                 const size_t index,
                                                 const std::string quantize_granularity)>;
 
+/*!
+ * \brief Register a function to determine if the input of a quantized operator
+ * needs to be asymmetric quantized.
```
[the remark this answers is not in the document]
Comment (bgawrych) [reply include/mxnet/op_attr_types.h]: done

File: src/operator/quantization/quantize_graph_pass.cc (hunk 1 above)
Comment (bgawrych) [reply src/operator/quantization/quantize_graph_pass.cc]: done

File: src/operator/quantization/mkldnn/mkldnn_quantized_rnn-inl.h (hunk 2 above)
Comment (bgawrych) [reply src/operator/quantization/mkldnn/mkldnn_quantized_rnn-inl.h]: done

File: src/operator/quantization/mkldnn/mkldnn_quantized_rnn.cc (hunk 3 above)
Comment (bgawrych) [reply src/operator/quantization/mkldnn/mkldnn_quantized_rnn.cc]: done

File: src/operator/quantization/quantize_asym.cc (hunk 5 above)
Comment (bgawrych) [reply src/operator/quantization/quantize_asym.cc]: done

File: src/operator/quantization/quantize_v2.cc (hunk 6 above)
Comment (bgawrych) [reply src/operator/quantization/quantize_v2.cc]: done

File: src/operator/quantization/quantized_rnn-inl.h (hunk 7 above)
Comment (bgawrych) [reply src/operator/quantization/quantized_rnn-inl.h]: done

File: src/operator/quantization/quantized_rnn.cc (hunk 8 above)
Comment (bgawrych) [reply src/operator/quantization/quantized_rnn.cc]: done

File: src/operator/quantization/quantize_asym-inl.h (hunk 4 above)
Comment (bgawrych) [reply src/operator/quantization/quantize_asym-inl.h]: done

File: src/operator/nn/mkldnn/mkldnn_rnn.cc [hunk 10]
```diff
@@ -586,6 +613,7 @@ void MKLDNNRnnForward::SetWeightsMem(MKLDNNRnnMemMgr* mgr,
     ConcatWeights(*weights_iter_r_, 1, {l2r_wh, r2l_wh}, format_tag::ldgoi);
   } else if (param_.num_layer == 1 && !param_.bidirectional) {
     //* single uni-directional layer, no concatenate operator needed
+    // tutttaj
```
Comment (bartekkuncer) [inline src/operator/nn/mkldnn/mkldnn_rnn.cc:616]: I believe this line of comment is unnecessary.

File: src/operator/nn/mkldnn/mkldnn_rnn.cc [hunk 11]
```diff
@@ -437,35 +472,27 @@ void MKLDNNRnnForward::SetNewDataMem(void* x,
...
+        net_args[MKLDNN_ARG_SRC] = src;
+        net_args[MKLDNN_ARG_DST] = dst;
+        MKLDNNStream::Get()->RegisterPrimArgs(mkldnn::reorder(reorder_pd), net_args);
+      };
+    ReorderWithAttr(*weights_layer_r_, *weights_layer_);
+    ReorderWithAttr(*weights_iter_r_, *weights_iter_);
+  } else {
+    MKLDNNMemoryReorder(*weights_layer_r_, *weights_layer_);
+    MKLDNNMemoryReorder(*weights_iter_r_, *weights_iter_);
+    // if (param_.proj_size > 0) MKLDNNMemoryReorder(*weights_proj_r_, *weights_proj_);
```
Comment (bartekkuncer) [inline src/operator/nn/mkldnn/mkldnn_rnn.cc:494]: Why is this here? Is it a TODO?

File: src/operator/nn/mkldnn/mkldnn_rnn.cc [hunk 12]
```diff
@@ -1047,6 +1075,12 @@ void MKLDNNRnnOp::Forward(const OpContext& ctx,
     weights_version_ = inputs[rnn_enum::kParams].version();
   }
 
+  if (dmlc::GetEnv("MXNET_RNN_USE_WEIGHT_CACHE", 0) && !initialized_) {
+    LOG(INFO) << "The current weight of RNN is assumed to be fixed and cached during "
+        "the whole inference pipeline. Please set MXNET_RNN_USE_WEIGHT_CACHE=0, if "
+        "the weight changed at runtime.";
```
Comment (bartekkuncer) [inline src/operator/nn/mkldnn/mkldnn_rnn.cc:1081]: "... is going to be changed (modified) at runtime." ?

File: src/operator/quantization/mkldnn/mkldnn_quantized_rnn.cc [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+    const int weight_nblks = layer_param.num_layer * directions;
+    for (int blk = 0; blk < weight_nblks; ++blk) {
+#pragma omp parallel for num_threads(nthreads)
+      for (index_t i = 0; i < static_cast<index_t>(wh_size0); ++i) {
+        goi_max[i] = MaxAbs(wx[i], wh[i]);
+      }
+      for (index_t go = 0; go < gates_nblks; ++go) {
+        float tmp = w_max[go];
+//* NOTES: min/max reductions were supported since OpenMP 3.1, which was
+// released in
```
Comment (bartekkuncer) [inline src/operator/quantization/mkldnn/mkldnn_quantized_rnn.cc:98]: Fix this comment please.

File: src/operator/quantization/quantize_graph_pass.cc [hunk 14]
```diff
@@ -249,8 +248,12 @@ static void MarkQuantizedNodes(const Graph& src,
 Graph QuantizeGraph(Graph &&src) {
   static const auto& flist_outputs = nnvm::Op::GetAttr<nnvm::FListOutputNames>("FListOutputNames");
   static const auto& need_requantize_map = Op::GetAttr<mxnet::FNeedRequantize>("FNeedRequantize");
```
Comment (bartekkuncer) [inline src/operator/quantization/quantize_graph_pass.cc:250]: Maybe clang format whole file? Or at least the lines around the code you changed/added?

**Review by bartekkuncer — CHANGES_REQUESTED** [review]

File: src/operator/nn/mkldnn/mkldnn_rnn.cc (hunk 11 above)
Comment (bgawrych) [reply src/operator/nn/mkldnn/mkldnn_rnn.cc]: Nope, projection was not added in 1.x - this can be addressed as other issue

File: src/operator/quantization/quantize_graph_pass.cc (hunk 14 above)
Comment (bgawrych) [reply src/operator/quantization/quantize_graph_pass.cc:250]: I didn't want to format whole files to keep change clean and easy reviewable - I have applied formatter to my changes

**Comment by bgawrych** [comment]
@josephevans @bartekkuncer @anko-intel Can you take a quick look again and check if your comments was resolved?

**Review by josephevans — APPROVED** [review]
Thanks for the updates, looks good.

**Review by anko-intel — APPROVED** [review]

**Review by bartekkuncer — APPROVED** [review]
